### PR TITLE
Fix warning: `comparison of integer expressions of different signedness`

### PR DESCRIPTION
**Pull request by SergeyKopienko**
/oneDPL/include/oneapi/dpl/pstl/./omp/parallel_scan.h:113:13: warning: comparison of integer expressions of different signedness: ‘long int’ and ‘const std::size_t’ {aka ‘const long unsigned int’} [-Wsign-compare]

This warning has been detected at https://github.com/uxlfoundation/oneDPL/actions/runs/28441982213?pr=2736

## Description
Several internal call sites compared integer expressions of different signedness by force-casting one operand (e.g. `static_cast<_Size>(__last - __first))`. Besides triggering `-Wsign-compare` warnings, this pattern is unsafe: when a signed difference is negative, casting it to an unsigned type produces a large positive value and yields an incorrect comparison result.
This PR introduces `C++17`-compatible helpers that replicate the semantics of `C++20`'s `std::cmp_less` / `std::cmp_less_equal`, and uses them at the affected call sites. This removes the warnings and fixes the latent incorrect-comparison behavior for negative signed values.

## Changes
include/oneapi/dpl/pstl/utils.h — add two helpers in oneapi::dpl::__internal:
•	`__cmp_less(T, U)`
•	`__cmp_less_equal(T, U)`
Both are `constexpr`, constrained to integral types via `std::enable_if_t`, and perform sign-aware comparisons without any narrowing/lossy casts.
Replace unsafe casted comparisons with the new helpers:
•	pstl/algorithm_impl.h (`__find_subrange`): two comparisons of `__global_last - __first` against `__count`.
•	pstl/omp/parallel_scan.h (`__parallel_strict_scan`): `__n` vs `__default_chunk_size`.
•	pstl/parallel_backend_tbb.h (`__parallel_merge`): `__n` vs `__merge_cut_off`.

## Notes
•	`C++17` baseline is preserved; no `C++20` library features (`std::cmp_*`) are required.
•	All replacements are semantically equivalent to the original intended comparisons (with argument order adjusted where needed, e.g. `count <= size` for the `>=` case).
•	Behavior of serial/parallel/device code paths is unchanged aside from correcting the negative-value edge case.

## Testing
•	Existing tests continue to pass.

### CODE REVIEW

File: include/oneapi/dpl/pstl/algorithm_impl.h [hunk 1]
```diff
@@ -25,6 +25,7 @@
 #include <tuple>
 #include <optional> // for std::optional
 #include <vector>
+#include <type_traits> // for std::common_type_t
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/algorithm_impl.h:28]: Fixed

File: include/oneapi/dpl/pstl/omp/parallel_scan.h [hunk 2]
```diff
@@ -110,7 +112,8 @@ void
 __parallel_strict_scan(oneapi::dpl::__internal::__omp_backend_tag, _ExecutionPolicy&&, _Index __n, _Tp __initial,
                        _Rp __reduce, _Cp __combine, _Sp __scan, _Ap __apex)
 {
-    if (__n <= __default_chunk_size)
+    using _CommonType = std::common_type_t<_Index, decltype(__default_chunk_size)>;
+    if (static_cast<_CommonType>(__n) <= static_cast<_CommonType>(__default_chunk_size))
```
Comment (akukanov) [inline include/oneapi/dpl/pstl/omp/parallel_scan.h]: `__default_chunk_size` is `std::size_t`, and we know that. I would just use `std::size_t(__n)`.

**Review by MikeDvorskiy — COMMENTED** [review]
Honestly speaking, It's getting a bit cumbersome....
Probably, does it make sense to change `size_t `to `int `for these constants? 

`constexpr int  __default_chunk_size = 2048;`
`constexpr int  __merge_cut_off = 2000;`

**Comment by SergeyKopienko** [comment]
> Honestly speaking, It's getting a bit cumbersome.... Probably, does it make sense to change `size_t `to `int `for these constants?
> 
> `constexpr int __default_chunk_size = 2048;` `constexpr int __merge_cut_off = 2000;`

I think in this case highly likely we will have another subset of warning.

The example:
```cpp
template <typename _RandomAccessIterator1, typename _RandomAccessIterator2, typename _RandomAccessIterator3,
          typename _Compare, typename _LeafMerge>
void
__parallel_merge_body(std::size_t __size_x, std::size_t __size_y, _RandomAccessIterator1 __xs,
                      _RandomAccessIterator1 __xe, _RandomAccessIterator2 __ys, _RandomAccessIterator2 __ye,
                      _RandomAccessIterator3 __zs, _Compare __comp, _LeafMerge __leaf_merge)
{

    if (__size_x + __size_y <= oneapi::dpl::__omp_backend::__default_chunk_size)  // <<<<<<

```

File: include/oneapi/dpl/pstl/algorithm_impl.h [hunk 3]
```diff
@@ -764,14 +765,17 @@ __find_subrange(_RandomAccessIterator __first, _RandomAccessIterator __last, _Ra
         return __last;
     }
 
+    using _CommonType = std::common_type_t<_Size, decltype(__global_last - __first)>;
+
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/algorithm_impl.h]: Fixed

File: include/oneapi/dpl/pstl/parallel_backend_tbb.h [hunk 4]
```diff
@@ -1259,7 +1259,9 @@ __parallel_merge(oneapi::dpl::__internal::__tbb_backend_tag, _ExecutionPolicy&&,
     using _DifferenceType2 = typename std::iterator_traits<_RandomAccessIterator2>::difference_type;
     using _SizeType = typename std::common_type_t<_DifferenceType1, _DifferenceType2>;
     const _SizeType __n = (__xe - __xs) + (__ye - __ys);
-    if (__n <= __merge_cut_off)
+
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/parallel_backend_tbb.h]: It makes sense to add
`assert(__n > 0);`

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/parallel_backend_tbb.h]: It's orthogonal to this PR
UPD: fixed

File: include/oneapi/dpl/pstl/omp/parallel_scan.h (hunk 2 above)
Comment (akukanov) [inline include/oneapi/dpl/pstl/omp/parallel_scan.h]: I think that taking common type is excessive in the cases when a) one of the types is well known (here, the type of `__default_chunk_size`) and b) it is known to be sufficient to hold the second value. Just cast to the known type.

File: include/oneapi/dpl/pstl/parallel_backend_tbb.h [hunk 5]
```diff
@@ -1259,7 +1259,9 @@ __parallel_merge(oneapi::dpl::__internal::__tbb_backend_tag, _ExecutionPolicy&&,
...
+    using _CommonType = std::common_type_t<_SizeType, decltype(__merge_cut_off)>;
+    if (static_cast<_CommonType>(__n) <= static_cast<_CommonType>(__merge_cut_off))
```
Comment (akukanov) [inline include/oneapi/dpl/pstl/parallel_backend_tbb.h]: Similar story here, we know and control the type of `__merge_cut_off`.

File: include/oneapi/dpl/pstl/algorithm_impl.h [hunk 6]
```diff
@@ -764,14 +764,17 @@ __find_subrange(_RandomAccessIterator __first, _RandomAccessIterator __last, _Ra
...
     auto __unary_pred =
         [__pred, &__value](auto&& __val) mutable { return __pred(std::forward<decltype(__val)>(__val), __value); };
-    while (__first != __last && (static_cast<_Size>(__global_last - __first) >= __count))
+    while (__first != __last && static_cast<_CommonType>(__global_last - __first) >= static_cast<_CommonType>(__count))
```
Comment (akukanov) [inline include/oneapi/dpl/pstl/algorithm_impl.h]: In this function, `global_last - first` is compared with `count` three times. That means, six casts to the common type. A much better way is to compute `global_last - count` once, and then compare the result with `first` (or their difference with 0).

**Comment by MikeDvorskiy** [comment]
> I think in this case highly likely we will have another subset of warning.

In case of positive constants (the mentioned constants are positive) should not be such warning.

**Review by dmitriy-sobolev — COMMENTED** [review]
Does the PR fix all current issues of this kind? If yes, we should remove  `-Wno-error=sign-compare` exclusion from `ci-testing.yml` and its `cl` equivalent.

**Comment by SergeyKopienko** [comment]
> Does the PR fixes all current issues of this kind? If yes, we should remove `-Wno-error=sign-compare` exclusion from `ci-testing.yml` and a `cl` equivalent.

This PR trying to fix all warnings of this kind which were detected for now.
But I think have sense to still use  `-Wno-error=sign-compare` option in the future too.

**Comment by dmitriy-sobolev** [comment]
> But I think have sense to still use `-Wno-error=sign-compare` option in the future too.

Why? The removal would prevent reintroduction of the same issues. However, we should keep the exclusions if e.g. SYCL or TBB have these warnings, which I doubt.

**Comment by SergeyKopienko** [comment]
> > But I think have sense to still use `-Wno-error=sign-compare` option in the future too.
> 
> Why? The removal would prevent reintroduction of the same issues. However, we should keep the exclusions if e.g. SYCL or TBB have these warnings, which I doubt.

...because we always can write some new code with the same problem.

File: include/oneapi/dpl/pstl/utils.h [hunk 7]
```diff
@@ -668,6 +668,31 @@ __dpl_signbit(const _T& __x)
...
+constexpr std::enable_if_t<std::is_integral_v<T> && std::is_integral_v<U>, bool>
+__cmp_less(T __t, U __u)
+{
+    if constexpr (std::is_signed_v<T> == std::is_signed_v<U>)
+        return __t < __u;
+    else if constexpr (std::is_signed_v<T>)
+        return __t < 0 || static_cast<std::make_unsigned_t<T>>(__t) < __u;
+    else
+        return __u >= 0 && __t < static_cast<std::make_unsigned_t<U>>(__u);
+}
```
Comment (akukanov) [inline include/oneapi/dpl/pstl/utils.h]: That's how I would do it:
```suggestion
__cmp_less(T __t, U __u)
{
    assert( __t > 0 && __u > 0);
    if constexpr (std::is_signed_v<T> == std::is_signed_v<U>)
        return __t < __u;
    else
        return static_cast<std::make_unsigned_t<T>>(__t) < static_cast<std::make_unsigned_t<U>>(__u);
}
```
because in all the cases the real arguments are not supposed to be negative, as far as I can see, and we do not want  extra checks that may not always be optimized out.

File: include/oneapi/dpl/pstl/utils.h [hunk 8]
```diff
@@ -668,6 +668,31 @@ __dpl_signbit(const _T& __x)
     return (__x & __mask) != 0;
 }
 
+// This prevents warning about comparison of integer expressions of different signedness
+template <typename T, typename U>
```
Comment (akukanov) [inline include/oneapi/dpl/pstl/utils.h]: The god of underscores demands more underscores here.

File: include/oneapi/dpl/pstl/utils.h (hunk 7 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/utils.h]: Thanks for the suggestion.
I'd prefer to keep the current implementation, for a few reasons:

1. It intentionally mirrors the reference implementation. The current `__cmp_less` / `__cmp_less_equal` follow the example implementation of `std::cmp_less` from cppreference (https://en.cppreference.com/w/cpp/utility/intcmp). Since our C++17 baseline can't use `std::cmp_less` directly, reproducing the standard, well-known behavior keeps the helper predictable and correct for all integer inputs, and easy to reason about for future readers.
2. The proposed `assert(__t > 0 && __u > 0)` is too strict. The condition should be `>= 0`, not `> 0`. Zero is a valid, frequently-occurring value at the call sites — e.g. `__global_last - __first`, `__n`, and `__merge_cut_off` can all be 0 for empty ranges. With `> 0` the assertion would fire in debug builds on perfectly valid inputs.
3. Replacing the logic weakens correctness in release. assert is compiled out in release builds. With the current code, if a negative value ever reaches the helper, the result is still correct. With the proposed version, casting a negative value to unsigned silently produces a huge number and a wrong result. Given that the whole point of this PR is to fix mismatched-signedness comparisons, I'd rather not introduce a helper named `__cmp_less` that silently diverges from std::cmp_less semantics.
4. The optimization concern doesn't really apply here. The signedness dispatch is already compile-time (`if constexpr`), so only one branch is generated. The remaining runtime check (`__t < 0` / `__u >= 0`) is a single comparison that the compiler typically folds away. All current uses are host-side control code (TBB merge cut-offs, search_n), not hot device kernels, so there's no measurable benefit and no register-pressure/64-bit-arithmetic concern to justify deviating from the standard behavior.
Suggested alternative: if we want to document the "non-negative" precondition, I'm happy to add assert(`__t >= 0 && __u >= 0`) on top of the existing correct logic (not replacing it). That keeps both the invariant documentation and release-build safety.

File: include/oneapi/dpl/pstl/utils.h (hunk 8 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/utils.h]: Fixed

File: include/oneapi/dpl/pstl/utils.h (hunk 7 above)
Comment (akukanov) [reply include/oneapi/dpl/pstl/utils.h]: The last thing I want is to argue with AI generated comments - I have more interesting things to do in my life, May I get a personal opinion instead, please?

**Comment by SergeyKopienko** [comment]
Closed as not actial
